kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -269,6 +277,19 @@ public Optional<SetTransaction> getSetTxnOpt() {
     return setTxnOpt;
   }
 
+  /**
+   * Add domain metadata to the transaction to be committed.
+   *
+   * @param domainMetadatas List of domain metadata to be added.
+   */
+  public void addDomainMetadata(List<DomainMetadata> domainMetadatas) {
```

Comment:
also -- this should be in the TransactionBuilderImpl class instead. is there a specific reason you want to `build` a Transaction and then mutate and append domain metadatas to it afterwards?

Reply:
I think a txn can contains multiple domain metadata. It seems to be the case in DBR. 

One reason I do the duplication check just before committing is that domain metadata might be mutated in other code paths, such as during conflict resolution. Validating them right before the commit helps ensure that we capture all changes.


> Is there a specific reason you want to build a Transaction and then mutate and append domain metadatas to it afterwards?

I'm not sure about other domain metadata usages. For Row Tracking, its domain metadata depends on all `AddFile` actions in the `dataActions` passed into `Transaction.commit()`. We can only append this domain metadata after building the transaction and seeing all `AddFile` actions. Also, conflict resolution can mutate existing domain metadatas.

Reply:
This error is currently thrown in two cases:
- When we detect duplicate domain metadata entries in a winning commit file during conflict resolution, which corresponds to the scenario you mentioned (read path).
- When validating domain metadata actions to be committed in `Transaction.commit()` (write path).

Perhaps we could have separate errors for each scenario. For instance, `commitDuplicateDomainMetadataAction` could throw a `KernelException`, and `tableWithDuplicateDomainMetadataAction` could throw an `InvalidTableException`.

Reply:
It seems like in practice we might not use an API like this? The domain metadata might just be generated as a __step__ during `commit`?

Reply:
I feel like I'm having a bit of a hard time mentally modeling this without an actual feature using domain metadatas. Are we adding support for row tracking or anything else after this?

If these are all private APIs we can revisit whether they make sense when adding support for a feature that uses them.

Reply:
> It seems like in practice we might not use an API like this? The domain metadata might just be generated as a step during commit?

Yes, currently this API is mainly used for tests. In actual use, I imagine domain metadata will be constructed during the commit process by feature-related code.

We will add support for row tracking after this PR. That would provide a clearer example of how domain metadata is used in practice. At that point, we may also revisit and adjust some APIs. For now, domain metadata is only used internally for table features, so these will remain private APIs.

Reply:
If this is used only for tests, can you please add `Visible for testing` to the javadoc.

Also -- @allisonport-db and @vkorukanti what do you think about adding a `@VisibleForTesting` annotation that we can add to methods to make tthis even clearer?

Reply:
+1 to documenting it's for testing (at least for now as long as no feature is using it)

Reply:
Annotation SGTM

Reply:
If it is only for testing now I think there's also some validity to moving it to `TransactionBuilderImpl` as @scottsand-db suggested to avoid mutating a transaction. But since it's just for testing maybe it's not that important ¯\_(ツ)_/¯

Reply:
I think avoiding mutating a transaction makes sense. I've moved it to `TransactionBuilderImpl` and mentioned `Visible for testing` in its javadoc.